In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

In [2]:
%matplotlib inline
pd.set_option('display.max_columns', None)

In [5]:
# import excel file November2012.xlsx
tn_2010_g=pd.read_excel('../data/2010_election/November2010.xls',
                      usecols=['COUNTY','Precinct Name','OFFICENAME','ELECTDATE','BName1','Tally1','BName2','Tally2',
                               'BName3','Tally3','BName4','Tally4','BName5','Tally5','BName6','Tally6','BName7','Tally7',
                               'BName8','Tally8','BName9','Tally9','BName10','Tally10'])
tn_2010_g

COUNTY  Precinct Name                                   OFFICENAME  \
0      ANDERSON  Andersonville                                     Governor   
1      ANDERSON  Andersonville                                     Governor   
2      ANDERSON  Andersonville                     Constitutional Amendment   
3      ANDERSON  Andersonville      U. S. House of Representatives District   
4      ANDERSON  Andersonville                    Tennessee Senate District   
...         ...            ...                                          ...   
12492    WILSON  25 Maple Hill                                     Governor   
12493    WILSON  25 Maple Hill                     Constitutional Amendment   
12494    WILSON  25 Maple Hill      U. S. House of Representatives District   
12495    WILSON  25 Maple Hill                    Tennessee Senate District   
12496    WILSON  25 Maple Hill  Tennessee House of Representatives District   

              ELECTDATE                    BName1  Tally1  \
0      November 2, 2010         . Bill Haslam - R     717   
1      November 2, 2010  . Donald Ray McFolin - I       0   
2      November 2, 2010                     . Yes     813   
3      November 2, 2010   . Chuck Fleischmann - R     519   
4      November 2, 2010       . Randy McNally - R     742   
...                 ...                       ...     ...   
12492  November 2, 2010  . Donald Ray McFolin - I       0   
12493  November 2, 2010                     . Yes    1207   
12494  November 2, 2010          . David Hall - R     963   
12495  November 2, 2010         . Mae Beavers - R     979   
12496  November 2, 2010          . Linda Elam - R     863   

                      BName2  Tally2                     BName3  Tally3  \
0       . Mike McWherter - D     164       . Bayron Binkley - I       1   
1      . Linda Kay Perry - I       0         . James Reesor - I       3   
2                       . No      46                        NaN       0   
3           . John Wolfe - D     141          . Don Barkman - I       2   
4                        NaN       0                        NaN       0   
...                      ...     ...                        ...     ...   
12492  . Linda Kay Perry - I       1         . James Reesor - I       0   
12493                   . No     108                        NaN       0   
12494       . Jim Cooper - D     408  . Stephen W. Collings - I       4   
12495  . George McDonald - D     398                        NaN       0   
12496      . Allen Barry - D     295    . Luther C. Lenning - I      92   

                      BName4  Tally4                    BName5  Tally5  \
0        . Brandon Dodds - I       8   . Samuel David Duck - I       0   
1      . Thomas Smith II - I       2   . Howard M. Switzer - I       1   
2                        NaN       0                       NaN       0   
3           . Mark DeVol - I     190  . Gregory C. Goodwin - I       3   
4                        NaN       0                       NaN       0   
...                      ...     ...                       ...     ...   
12492  . Thomas Smith II - I       1   . Howard M. Switzer - I       2   
12493                    NaN       0                       NaN       0   
12494       . Bill Crook - I       2   . John P. Miglietta - I       1   
12495                    NaN       0                       NaN       0   
12496    . Heather Scott - I     110                       NaN       0   

                                BName6  Tally6                  BName7  \
0                 . David Gatchell - I       0      . June Griffin - I   
1      . Carl Twofeathers Whitaker - I       8  . Write-in - James Lee   
2                                  NaN       0                     NaN   
3               . Robert Humphries - I       1           . Mo Kiah - I   
4                                  NaN       0                     NaN   
...                                ...     ...                     ...   
12492  . Carl Twofe

In [6]:
tn_2010_g.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12497 entries, 0 to 12496
Data columns (total 24 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   COUNTY         12497 non-null  object
 1   Precinct Name  12497 non-null  object
 2   OFFICENAME     12497 non-null  object
 3   ELECTDATE      12497 non-null  object
 4   BName1         12497 non-null  object
 5   Tally1         12497 non-null  int64 
 6   BName2         11339 non-null  object
 7   Tally2         12497 non-null  int64 
 8   BName3         7213 non-null   object
 9   Tally3         12497 non-null  int64 
 10  BName4         6483 non-null   object
 11  Tally4         12497 non-null  int64 
 12  BName5         5992 non-null   object
 13  Tally5         12497 non-null  int64 
 14  BName6         5716 non-null   object
 15  Tally6         12497 non-null  int64 
 16  BName7         5163 non-null   object
 17  Tally7         12497 non-null  int64 
 18  BName8         4891 non-nu

In [10]:
# merge entries in all RNAME columns into new column "name"
tn_2010_g_namemelt=tn_2010_g.melt(id_vars=['COUNTY','Precinct Name','OFFICENAME','ELECTDATE'],value_vars=['BName1','BName2','BName3','BName4','BName5','BName6','BName7','BName8','BName9','BName10'],value_name="candidate")
# drop column "variable" created by .melt
tn_2010_g_namemelt=tn_2010_g_namemelt.drop(['variable'],axis=1)                          
tn_2010_g_namemelt.head()

COUNTY  Precinct Name                               OFFICENAME  \
0  ANDERSON  Andersonville                                 Governor   
1  ANDERSON  Andersonville                                 Governor   
2  ANDERSON  Andersonville                 Constitutional Amendment   
3  ANDERSON  Andersonville  U. S. House of Representatives District   
4  ANDERSON  Andersonville                Tennessee Senate District   

          ELECTDATE                 candidate  
0  November 2, 2010         . Bill Haslam - R  
1  November 2, 2010  . Donald Ray McFolin - I  
2  November 2, 2010                     . Yes  
3  November 2, 2010   . Chuck Fleischmann - R  
4  November 2, 2010       . Randy McNally - R

In [11]:
# merge entries in all TALLY columns into new column "vote_count"
tn_2010_g_pvtallymelt=tn_2010_g.melt(id_vars=['COUNTY','Precinct Name','OFFICENAME','ELECTDATE'],value_vars=['Tally1','Tally2','Tally3','Tally4','Tally5','Tally6','Tally7','Tally8','Tally9','Tally10'],value_name="vote_count")
# drop column "variable" created by .melt
tn_2010_g_pvtallymelt=tn_2010_g_pvtallymelt.drop(['variable'],axis=1)
tn_2010_g_pvtallymelt.head()

COUNTY  Precinct Name                               OFFICENAME  \
0  ANDERSON  Andersonville                                 Governor   
1  ANDERSON  Andersonville                                 Governor   
2  ANDERSON  Andersonville                 Constitutional Amendment   
3  ANDERSON  Andersonville  U. S. House of Representatives District   
4  ANDERSON  Andersonville                Tennessee Senate District   

          ELECTDATE  vote_count  
0  November 2, 2010         717  
1  November 2, 2010           0  
2  November 2, 2010         813  
3  November 2, 2010         519  
4  November 2, 2010         742

In [13]:
# add "vote_count" to tn_2016_namemelt
tally=tn_2010_g_pvtallymelt['vote_count']
tn_2010_g_namemelt=tn_2010_g_namemelt.join(tally)
tn_2010_g_namemelt.head(3)

COUNTY  Precinct Name                OFFICENAME         ELECTDATE  \
0  ANDERSON  Andersonville                  Governor  November 2, 2010   
1  ANDERSON  Andersonville                  Governor  November 2, 2010   
2  ANDERSON  Andersonville  Constitutional Amendment  November 2, 2010   

                  candidate  vote_count  
0         . Bill Haslam - R         717  
1  . Donald Ray McFolin - I           0  
2                     . Yes         813

In [14]:
tn_2010_g_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124970 entries, 0 to 124969
Data columns (total 6 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   COUNTY         124970 non-null  object
 1   Precinct Name  124970 non-null  object
 2   OFFICENAME     124970 non-null  object
 3   ELECTDATE      124970 non-null  object
 4   candidate      66327 non-null   object
 5   vote_count     124970 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 5.7+ MB


In [15]:
# lower case column names
tn_2010_g_namemelt.columns=tn_2010_g_namemelt.columns.str.lower()
tn_2010_g_namemelt.head(2)

county  precinct name officename         electdate  \
0  ANDERSON  Andersonville   Governor  November 2, 2010   
1  ANDERSON  Andersonville   Governor  November 2, 2010   

                  candidate  vote_count  
0         . Bill Haslam - R         717  
1  . Donald Ray McFolin - I           0

In [16]:
tn_2010_g_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124970 entries, 0 to 124969
Data columns (total 6 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   county         124970 non-null  object
 1   precinct name  124970 non-null  object
 2   officename     124970 non-null  object
 3   electdate      124970 non-null  object
 4   candidate      66327 non-null   object
 5   vote_count     124970 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 5.7+ MB


#### 66327 non-null names  out of 124970 enties. Confirm all are 0 votes and drop rows with no candidate listed

In [19]:
# Create dataframe "blank_candidate_2010" of null valuea in "candidate" of tn_2010_g_namemelt
blank_candidate_2010=tn_2010_g_namemelt[tn_2010_g_namemelt['candidate'].isna()]
blank_candidate_2010

county      precinct name  \
12501   ANDERSON      Andersonville   
12507   ANDERSON         Briceville   
12513   ANDERSON  Blockhouse Valley   
12519   ANDERSON            Clinton   
12525   ANDERSON       Clinton High   
...          ...                ...   
124963    WILSON        25 Leeville   
124965    WILSON      25 Maple Hill   
124966    WILSON      25 Maple Hill   
124968    WILSON      25 Maple Hill   
124969    WILSON      25 Maple Hill   

                                         officename         electdate  \
12501                     Tennessee Senate District  November 2, 2010   
12507                     Tennessee Senate District  November 2, 2010   
12513                     Tennessee Senate District  November 2, 2010   
12519                     Tennessee Senate District  November 2, 2010   
12525                     Tennessee Senate District  November 2, 2010   
...                                             ...               ...   
124963  Tennessee House of Representatives District  November 2, 2010   
124965                                     Governor  November 2, 2010   
124966                     Constitutional Amendment  November 2, 2010   
124968                    Tennessee Senate District  November 2, 2010   
124969  Tennessee House of Representatives District  November 2, 2010   

       candidate  vote_count  
12501        NaN           0  
12507        NaN           0  
12513        NaN           0  
12519        NaN           0  
12525        NaN           0  
...          ...         ...  
124963       NaN           0  
124965       NaN           0  
124966       NaN           0  
124968       NaN           0  
124969       NaN           0  

[58643 rows x 6 columns]

In [20]:
# confirm all are 0 counts
blank_candidate_2010['vote_count'].value_counts()

0    58643
Name: vote_count, dtype: int64

In [27]:
# drop all null candidates
tn_2010_g_namemelt.dropna().reset_index(drop=True)

county  precinct name                               officename  \
0      ANDERSON  Andersonville                                 Governor   
1      ANDERSON  Andersonville                                 Governor   
2      ANDERSON  Andersonville                 Constitutional Amendment   
3      ANDERSON  Andersonville  U. S. House of Representatives District   
4      ANDERSON  Andersonville                Tennessee Senate District   
...         ...            ...                                      ...   
66322    WILSON   24 First UMC  U. S. House of Representatives District   
66323    WILSON    25 Leeville                                 Governor   
66324    WILSON    25 Leeville  U. S. House of Representatives District   
66325    WILSON  25 Maple Hill                                 Governor   
66326    WILSON  25 Maple Hill  U. S. House of Representatives District   

              electdate                    candidate  vote_count  
0      November 2, 2010            . Bill Haslam - R         717  
1      November 2, 2010     . Donald Ray McFolin - I           0  
2      November 2, 2010                        . Yes         813  
3      November 2, 2010      . Chuck Fleischmann - R         519  
4      November 2, 2010          . Randy McNally - R         742  
...                 ...                          ...         ...  
66322  November 2, 2010  . James G. Whitfield II - I           3  
66323  November 2, 2010        . Boyce T. McCall - I           1  
66324  November 2, 2010  . James G. Whitfield II - I           2  
66325  November 2, 2010        . Boyce T. McCall - I           0  
66326  November 2, 2010  . James G. Whitfield II - I           4  

[66327 rows x 6 columns]

In [28]:
# confirn all null candidates dropped
tn_2010_g_namemelt[tn_2010_g_namemelt['candidate'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 58643 entries, 12501 to 124969
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   county         58643 non-null  object
 1   precinct name  58643 non-null  object
 2   officename     58643 non-null  object
 3   electdate      58643 non-null  object
 4   candidate      0 non-null      object
 5   vote_count     58643 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 3.1+ MB


In [24]:
tn_2010_g_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124970 entries, 0 to 124969
Data columns (total 6 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   county         124970 non-null  object
 1   precinct name  124970 non-null  object
 2   officename     124970 non-null  object
 3   electdate      124970 non-null  object
 4   candidate      66327 non-null   object
 5   vote_count     124970 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 5.7+ MB


### Look at cadidates with 0 votes, drop

In [29]:
tn_2010_g_namemelt_0votes=tn_2010_g_namemelt[tn_2010_g_namemelt['vote_count'] == 0]
tn_2010_g_namemelt_0votes

county      precinct name  \
1       ANDERSON      Andersonville   
7       ANDERSON         Briceville   
13      ANDERSON  Blockhouse Valley   
19      ANDERSON            Clinton   
25      ANDERSON       Clinton High   
...          ...                ...   
124964    WILSON      25 Maple Hill   
124965    WILSON      25 Maple Hill   
124966    WILSON      25 Maple Hill   
124968    WILSON      25 Maple Hill   
124969    WILSON      25 Maple Hill   

                                         officename         electdate  \
1                                          Governor  November 2, 2010   
7                                          Governor  November 2, 2010   
13                                         Governor  November 2, 2010   
19                                         Governor  November 2, 2010   
25                                         Governor  November 2, 2010   
...                                             ...               ...   
124964                                     Governor  November 2, 2010   
124965                                     Governor  November 2, 2010   
124966                     Constitutional Amendment  November 2, 2010   
124968                    Tennessee Senate District  November 2, 2010   
124969  Tennessee House of Representatives District  November 2, 2010   

                       candidate  vote_count  
1       . Donald Ray McFolin - I           0  
7       . Donald Ray McFolin - I           0  
13      . Donald Ray McFolin - I           0  
19      . Donald Ray McFolin - I           0  
25      . Donald Ray McFolin - I           0  
...                          ...         ...  
124964     . Boyce T. McCall - I           0  
124965                       NaN           0  
124966                       NaN           0  
124968                       NaN           0  
124969                       NaN           0  

[85394 rows x 6 columns]

In [21]:
# drop rows with 0 in "vote_count" column
tn_2012_namemelt=tn_2012_namemelt[tn_2012_namemelt['vote_count'] != 0].reset_index(drop=True)
tn_2012_namemelt

county       precinct                                   officename  \
0        ANDERSON  Andersonville                      United States President   
1        ANDERSON  Andersonville                         United States Senate   
2        ANDERSON  Andersonville       U.S. House of Representatives District   
3        ANDERSON  Andersonville  Tennessee House of Representatives District   
4        ANDERSON     Briceville                      United States President   
...           ...            ...                                          ...   
43415  WILLIAMSON           12-1                         United States Senate   
43416  WILLIAMSON           12-2                         United States Senate   
43417  WILLIAMSON           12-3                         United States Senate   
43418      WILSON           16-1                         United States Senate   
43419      WILSON           22-1                         United States Senate   

              electdate                      candidate  vote_count  
0      November 6, 2012            . Mitt Romney - (R)        1110  
1      November 6, 2012             . Bob Corker - (R)        1067  
2      November 6, 2012      . Chuck Fleischmann - (R)         996  
3      November 6, 2012          . Dennis Powers - (R)         998  
4      November 6, 2012            . Mitt Romney - (R)         227  
...                 ...                            ...         ...  
43415  November 6, 2012  . Write-In - Angelia Stinnett           2  
43416  November 6, 2012  . Write-In - Angelia Stinnett          13  
43417  November 6, 2012  . Write-In - Angelia Stinnett           3  
43418  November 6, 2012  . Write-In - Angelia Stinnett           1  
43419  November 6, 2012  . Write-In - Angelia Stinnett           1  

[43420 rows x 6 columns]

In [22]:
tn_2012_namemelt['vote_count'].value_counts()

1       4472
2       3531
3       2819
4       2166
5       1669
        ... 
1766       1
2611       1
1670       1
1638       1
4839       1
Name: vote_count, Length: 2260, dtype: int64

In [23]:
# group county precincts into county per candidate, in descending order of votes within county
tn_2012_int=tn_2012_namemelt.groupby(['electdate','officename','county','candidate'], as_index=False)['vote_count'].sum()
tn_2012_int=tn_2012_int.sort_values(by=['county','vote_count'],ascending=False).reset_index(drop=True)
tn_2012_int.head(30)

electdate                                   officename      county  \
0   November 6, 2012                         United States Senate      WILSON   
1   November 6, 2012                      United States President      WILSON   
2   November 6, 2012       U.S. House of Representatives District      WILSON   
3   November 6, 2012  Tennessee House of Representatives District      WILSON   
4   November 6, 2012                      United States President      WILSON   
5   November 6, 2012  Tennessee House of Representatives District      WILSON   
6   November 6, 2012                         United States Senate      WILSON   
7   November 6, 2012       U.S. House of Representatives District      WILSON   
8   November 6, 2012       U.S. House of Representatives District      WILSON   
9   November 6, 2012                         United States Senate      WILSON   
10  November 6, 2012                         United States Senate      WILSON   
11  November 6, 2012                         United States Senate      WILSON   
12  November 6, 2012                      United States President      WILSON   
13  November 6, 2012                         United States Senate      WILSON   
14  November 6, 2012                         United States Senate      WILSON   
15  November 6, 2012                         United States Senate      WILSON   
16  November 6, 2012                         United States Senate      WILSON   
17  November 6, 2012                      United States President      WILSON   
18  November 6, 2012                      United States President      WILSON   
19  November 6, 2012                      United States President      WILSON   
20  November 6, 2012                      United States President      WILSON   
21  November 6, 2012       U.S. House of Representatives District      WILSON   
22  November 6, 2012                         United States Senate      WILSON   
23  November 6, 2012                         United States Senate      WILSON   
24  November 6, 2012                         United States Senate      WILSON   
25  November 6, 2012                         United States Senate      WILSON   
26  November 6, 2012                         United States Senate  WILLIAMSON   
27  November 6, 2012       U.S. House of Representatives District  WILLIAMSON   
28  November 6, 2012                      United States President  WILLIAMSON   
29  November 6, 2012                      United States President  WILLIAMSON   

                             candidate  vote_count  
0                   . Bob Corker - (R)       36993  
1                  . Mitt Romney - (R)       36109  
2                  . Diane Black - (R)       35659  
3                   . Susan Lynn - (R)       24116  
4                 . Barack Obama - (D)       14695  
5                 . Mark A. Pody - (R)       12731  
6              . Mark E. Clayton - (D)       10497  
7                . Scott Beasley - (I)        7101  
8                    . Pat Riley - (G)        4474  
9              . Martin Pleasant - (G)         749  
10                . Kermit Steck - (C)         512  
11            . Shaun E. Crowell - (I)         450  
12                . Gary Johnson - (I)         381  
13        . Troy Stephen Scoggin - (I)         176  
14                . James Higdon - (I)         136  
15          . Michel Joseph Long - (I)         131  
16              . David Gatchell - (I)         121  
17                . Virgil Goode - (C)         117  
18                  . Jill Stein - (G)         113  
19    . Ross C. "Rocky" Anderson - (I)          43  
20               . Merlin Miller - (I)          41  
21        . Write-In - Rachel Robinson          13  
22           . Write-In - Jacob Maurer           7  
23             . Write-In - Larry Crim           7  
24  . Write-In - Christopher Schappert           3  
25       . Write-In - Angelia Stinnett           2  
26                  . Bob Corker - (R)       72402  
27            .

In [24]:
# Capitalize couty names
tn_2012_int['county']=tn_2012_int['county'].str.title()
tn_2012_int.head(5)

electdate                                   officename  county  \
0  November 6, 2012                         United States Senate  Wilson   
1  November 6, 2012                      United States President  Wilson   
2  November 6, 2012       U.S. House of Representatives District  Wilson   
3  November 6, 2012  Tennessee House of Representatives District  Wilson   
4  November 6, 2012                      United States President  Wilson   

              candidate  vote_count  
0    . Bob Corker - (R)       36993  
1   . Mitt Romney - (R)       36109  
2   . Diane Black - (R)       35659  
3    . Susan Lynn - (R)       24116  
4  . Barack Obama - (D)       14695

In [25]:
tn_2012_int.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2445 entries, 0 to 2444
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   electdate   2445 non-null   object
 1   officename  2445 non-null   object
 2   county      2445 non-null   object
 3   candidate   2445 non-null   object
 4   vote_count  2445 non-null   int64 
dtypes: int64(1), object(4)
memory usage: 95.6+ KB


In [26]:
tn_2012_int['officename'].unique()

array(['United States Senate', 'United States President',
       'U.S. House of Representatives District',
       'Tennessee House of Representatives District',
       'Tennessee Senate District'], dtype=object)

### Clean up names and greate separate party column

In [27]:
# new dataframe new_names of non-write-in candidates
new_names=tn_2012_int[tn_2012_int['candidate'].str.contains("Write")==False]
new_names

electdate                                   officename    county  \
0     November 6, 2012                         United States Senate    Wilson   
1     November 6, 2012                      United States President    Wilson   
2     November 6, 2012       U.S. House of Representatives District    Wilson   
3     November 6, 2012  Tennessee House of Representatives District    Wilson   
4     November 6, 2012                      United States President    Wilson   
...                ...                                          ...       ...   
2435  November 6, 2012                      United States President  Anderson   
2436  November 6, 2012                         United States Senate  Anderson   
2437  November 6, 2012                         United States Senate  Anderson   
2438  November 6, 2012                      United States President  Anderson   
2439  November 6, 2012                      United States President  Anderson   

                             candidate  vote_count  
0                   . Bob Corker - (R)       36993  
1                  . Mitt Romney - (R)       36109  
2                  . Diane Black - (R)       35659  
3                   . Susan Lynn - (R)       24116  
4                 . Barack Obama - (D)       14695  
...                                ...         ...  
2435                . Jill Stein - (G)          88  
2436      . Troy Stephen Scoggin - (I)          72  
2437            . David Gatchell - (I)          60  
2438  . Ross C. "Rocky" Anderson - (I)          54  
2439             . Merlin Miller - (I)          25  

[2311 rows x 5 columns]

In [28]:
#drop period in front of names
new_names['candidate']=new_names['candidate'].str[1:]
new_names

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


electdate                                   officename    county  \
0     November 6, 2012                         United States Senate    Wilson   
1     November 6, 2012                      United States President    Wilson   
2     November 6, 2012       U.S. House of Representatives District    Wilson   
3     November 6, 2012  Tennessee House of Representatives District    Wilson   
4     November 6, 2012                      United States President    Wilson   
...                ...                                          ...       ...   
2435  November 6, 2012                      United States President  Anderson   
2436  November 6, 2012                         United States Senate  Anderson   
2437  November 6, 2012                         United States Senate  Anderson   
2438  November 6, 2012                      United States President  Anderson   
2439  November 6, 2012                      United States President  Anderson   

                            candidate  vote_count  
0                    Bob Corker - (R)       36993  
1                   Mitt Romney - (R)       36109  
2                   Diane Black - (R)       35659  
3                    Susan Lynn - (R)       24116  
4                  Barack Obama - (D)       14695  
...                               ...         ...  
2435                 Jill Stein - (G)          88  
2436       Troy Stephen Scoggin - (I)          72  
2437             David Gatchell - (I)          60  
2438   Ross C. "Rocky" Anderson - (I)          54  
2439              Merlin Miller - (I)          25  

[2311 rows x 5 columns]

In [29]:
# split name fromm party desination
new_names=new_names['candidate'].str.split(pat='-',expand=True)
new_names

0     1
0                    Bob Corker    (R)
1                   Mitt Romney    (R)
2                   Diane Black    (R)
3                    Susan Lynn    (R)
4                  Barack Obama    (D)
...                          ...   ...
2435                 Jill Stein    (G)
2436       Troy Stephen Scoggin    (I)
2437             David Gatchell    (I)
2438   Ross C. "Rocky" Anderson    (I)
2439              Merlin Miller    (I)

[2311 rows x 2 columns]

In [30]:
new_names.rename(columns={0:'name',1:'party'},inplace=True)
new_names

name party
0                    Bob Corker    (R)
1                   Mitt Romney    (R)
2                   Diane Black    (R)
3                    Susan Lynn    (R)
4                  Barack Obama    (D)
...                          ...   ...
2435                 Jill Stein    (G)
2436       Troy Stephen Scoggin    (I)
2437             David Gatchell    (I)
2438   Ross C. "Rocky" Anderson    (I)
2439              Merlin Miller    (I)

[2311 rows x 2 columns]

In [31]:
# new dataframe tn_2012_int_wi of only write-in candidates
tn_2012_int_wi=tn_2012_int[tn_2012_int['candidate'].str.contains("\(")==False]
tn_2012_int_wi

electdate                              officename    county  \
21    November 6, 2012  U.S. House of Representatives District    Wilson   
22    November 6, 2012                    United States Senate    Wilson   
23    November 6, 2012                    United States Senate    Wilson   
24    November 6, 2012                    United States Senate    Wilson   
25    November 6, 2012                    United States Senate    Wilson   
...                ...                                     ...       ...   
2440  November 6, 2012  U.S. House of Representatives District  Anderson   
2441  November 6, 2012                    United States Senate  Anderson   
2442  November 6, 2012                    United States Senate  Anderson   
2443  November 6, 2012                    United States Senate  Anderson   
2444  November 6, 2012                    United States Senate  Anderson   

                               candidate  vote_count  
21          . Write-In - Rachel Robinson          13  
22             . Write-In - Jacob Maurer           7  
23               . Write-In - Larry Crim           7  
24    . Write-In - Christopher Schappert           3  
25         . Write-In - Angelia Stinnett           2  
...                                  ...         ...  
2440           . Write-In - Richard Hall           5  
2441           . Write-In - Jacob Maurer           5  
2442            . Write-In - Jim Maynard           2  
2443       . Write-In - Angelia Stinnett           1  
2444             . Write-In - Larry Crim           1  

[134 rows x 5 columns]

In [32]:
# split name fromm 'Write-in'
new_wi_name=tn_2012_int_wi['candidate'].str.split(pat='-',n=2,expand=True)
new_wi_name

0    1                       2
21    . Write  In          Rachel Robinson
22    . Write  In             Jacob Maurer
23    . Write  In               Larry Crim
24    . Write  In    Christopher Schappert
25    . Write  In         Angelia Stinnett
...       ...  ...                     ...
2440  . Write  In             Richard Hall
2441  . Write  In             Jacob Maurer
2442  . Write  In              Jim Maynard
2443  . Write  In         Angelia Stinnett
2444  . Write  In               Larry Crim

[134 rows x 3 columns]

In [33]:
# create new 'party' columnn with "Write-In"
new_wi_name['party']=new_wi_name[[0,1]].agg('-'.join,axis=1).str[2:]
new_wi_name

0    1                       2      party
21    . Write  In          Rachel Robinson  Write-In 
22    . Write  In             Jacob Maurer  Write-In 
23    . Write  In               Larry Crim  Write-In 
24    . Write  In    Christopher Schappert  Write-In 
25    . Write  In         Angelia Stinnett  Write-In 
...       ...  ...                     ...        ...
2440  . Write  In             Richard Hall  Write-In 
2441  . Write  In             Jacob Maurer  Write-In 
2442  . Write  In              Jim Maynard  Write-In 
2443  . Write  In         Angelia Stinnett  Write-In 
2444  . Write  In               Larry Crim  Write-In 

[134 rows x 4 columns]

In [34]:
#drop columns 0 and 1, rename column 2 to"candidate"
new_wi_name=new_wi_name.drop([0,1],axis=1)
new_wi_name

2      party
21           Rachel Robinson  Write-In 
22              Jacob Maurer  Write-In 
23                Larry Crim  Write-In 
24     Christopher Schappert  Write-In 
25          Angelia Stinnett  Write-In 
...                      ...        ...
2440            Richard Hall  Write-In 
2441            Jacob Maurer  Write-In 
2442             Jim Maynard  Write-In 
2443        Angelia Stinnett  Write-In 
2444              Larry Crim  Write-In 

[134 rows x 2 columns]

In [35]:
new_wi_name.rename(columns={2:'name'},inplace=True)
new_wi_name

name      party
21           Rachel Robinson  Write-In 
22              Jacob Maurer  Write-In 
23                Larry Crim  Write-In 
24     Christopher Schappert  Write-In 
25          Angelia Stinnett  Write-In 
...                      ...        ...
2440            Richard Hall  Write-In 
2441            Jacob Maurer  Write-In 
2442             Jim Maynard  Write-In 
2443        Angelia Stinnett  Write-In 
2444              Larry Crim  Write-In 

[134 rows x 2 columns]

In [36]:
# join new_names and new_wi_names
frames=[new_names,new_wi_name]
cleaned_names=pd.concat(frames)
cleaned_names

name      party
0           Bob Corker         (R)
1          Mitt Romney         (R)
2          Diane Black         (R)
3           Susan Lynn         (R)
4         Barack Obama         (D)
...                 ...        ...
2440       Richard Hall  Write-In 
2441       Jacob Maurer  Write-In 
2442        Jim Maynard  Write-In 
2443   Angelia Stinnett  Write-In 
2444         Larry Crim  Write-In 

[2445 rows x 2 columns]

### Join cleaned names and party column to tn_2012_int

In [37]:
name=cleaned_names['name']
party=cleaned_names['party']
tn_2012_final=tn_2012_int.join(name).join(party)
tn_2012_final

electdate                                   officename    county  \
0     November 6, 2012                         United States Senate    Wilson   
1     November 6, 2012                      United States President    Wilson   
2     November 6, 2012       U.S. House of Representatives District    Wilson   
3     November 6, 2012  Tennessee House of Representatives District    Wilson   
4     November 6, 2012                      United States President    Wilson   
...                ...                                          ...       ...   
2440  November 6, 2012       U.S. House of Representatives District  Anderson   
2441  November 6, 2012                         United States Senate  Anderson   
2442  November 6, 2012                         United States Senate  Anderson   
2443  November 6, 2012                         United States Senate  Anderson   
2444  November 6, 2012                         United States Senate  Anderson   

                          candidate  vote_count               name      party  
0                . Bob Corker - (R)       36993        Bob Corker         (R)  
1               . Mitt Romney - (R)       36109       Mitt Romney         (R)  
2               . Diane Black - (R)       35659       Diane Black         (R)  
3                . Susan Lynn - (R)       24116        Susan Lynn         (R)  
4              . Barack Obama - (D)       14695      Barack Obama         (D)  
...                             ...         ...                ...        ...  
2440      . Write-In - Richard Hall           5       Richard Hall  Write-In   
2441      . Write-In - Jacob Maurer           5       Jacob Maurer  Write-In   
2442       . Write-In - Jim Maynard           2        Jim Maynard  Write-In   
2443  . Write-In - Angelia Stinnett           1   Angelia Stinnett  Write-In   
2444        . Write-In - Larry Crim           1         Larry Crim  Write-In   

[2445 rows x 7 columns]

In [38]:
tn_2012_final['party'].value_counts()

 (I)         940
 (R)         504
 (D)         421
 (G)         256
 (C)         190
Write-In     134
Name: party, dtype: int64

In [39]:
# change "(R)" to "Republican" and "(D)"to "Democratic" in "party" column
tn_2012_final["party"]=tn_2012_final['party'].str.replace('\(R\)',"Republican").str.replace('\(D\)',"Democratic")
tn_2012_final

electdate                                   officename    county  \
0     November 6, 2012                         United States Senate    Wilson   
1     November 6, 2012                      United States President    Wilson   
2     November 6, 2012       U.S. House of Representatives District    Wilson   
3     November 6, 2012  Tennessee House of Representatives District    Wilson   
4     November 6, 2012                      United States President    Wilson   
...                ...                                          ...       ...   
2440  November 6, 2012       U.S. House of Representatives District  Anderson   
2441  November 6, 2012                         United States Senate  Anderson   
2442  November 6, 2012                         United States Senate  Anderson   
2443  November 6, 2012                         United States Senate  Anderson   
2444  November 6, 2012                         United States Senate  Anderson   

                          candidate  vote_count               name  \
0                . Bob Corker - (R)       36993        Bob Corker    
1               . Mitt Romney - (R)       36109       Mitt Romney    
2               . Diane Black - (R)       35659       Diane Black    
3                . Susan Lynn - (R)       24116        Susan Lynn    
4              . Barack Obama - (D)       14695      Barack Obama    
...                             ...         ...                ...   
2440      . Write-In - Richard Hall           5       Richard Hall   
2441      . Write-In - Jacob Maurer           5       Jacob Maurer   
2442       . Write-In - Jim Maynard           2        Jim Maynard   
2443  . Write-In - Angelia Stinnett           1   Angelia Stinnett   
2444        . Write-In - Larry Crim           1         Larry Crim   

            party  
0      Republican  
1      Republican  
2      Republican  
3      Republican  
4      Democratic  
...           ...  
2440    Write-In   
2441    Write-In   
2442    Write-In   
2443    Write-In   
2444    Write-In   

[2445 rows x 7 columns]

In [40]:
# create data frame tn_2012_pres for presidental election
tn_2012_pres=tn_2012_final[tn_2012_final['officename'].str.match('United States President')]
tn_2012_pres=tn_2012_pres.reset_index(drop=True)
tn_2012_pres

electdate               officename    county  \
0    November 6, 2012  United States President    Wilson   
1    November 6, 2012  United States President    Wilson   
2    November 6, 2012  United States President    Wilson   
3    November 6, 2012  United States President    Wilson   
4    November 6, 2012  United States President    Wilson   
..                ...                      ...       ...   
659  November 6, 2012  United States President  Anderson   
660  November 6, 2012  United States President  Anderson   
661  November 6, 2012  United States President  Anderson   
662  November 6, 2012  United States President  Anderson   
663  November 6, 2012  United States President  Anderson   

                            candidate  vote_count                        name  \
0                 . Mitt Romney - (R)       36109                Mitt Romney    
1                . Barack Obama - (D)       14695               Barack Obama    
2                . Gary Johnson - (I)         381               Gary Johnson    
3                . Virgil Goode - (C)         117               Virgil Goode    
4                  . Jill Stein - (G)         113                 Jill Stein    
..                                ...         ...                         ...   
659              . Gary Johnson - (I)         301               Gary Johnson    
660              . Virgil Goode - (C)         101               Virgil Goode    
661                . Jill Stein - (G)          88                 Jill Stein    
662  . Ross C. "Rocky" Anderson - (I)          54   Ross C. "Rocky" Anderson    
663             . Merlin Miller - (I)          25              Merlin Miller    

           party  
0     Republican  
1     Democratic  
2            (I)  
3            (C)  
4            (G)  
..           ...  
659          (I)  
660          (C)  
661          (G)  
662          (I)  
663          (I)  

[664 rows x 7 columns]

In [41]:
tn_2012_pres.head(20)

electdate               officename      county  \
0   November 6, 2012  United States President      Wilson   
1   November 6, 2012  United States President      Wilson   
2   November 6, 2012  United States President      Wilson   
3   November 6, 2012  United States President      Wilson   
4   November 6, 2012  United States President      Wilson   
5   November 6, 2012  United States President      Wilson   
6   November 6, 2012  United States President      Wilson   
7   November 6, 2012  United States President  Williamson   
8   November 6, 2012  United States President  Williamson   
9   November 6, 2012  United States President  Williamson   
10  November 6, 2012  United States President  Williamson   
11  November 6, 2012  United States President  Williamson   
12  November 6, 2012  United States President  Williamson   
13  November 6, 2012  United States President  Williamson   
14  November 6, 2012  United States President       White   
15  November 6, 2012  United States President       White   
16  November 6, 2012  United States President       White   
17  November 6, 2012  United States President       White   
18  November 6, 2012  United States President       White   
19  November 6, 2012  United States President       White   

                           candidate  vote_count                        name  \
0                . Mitt Romney - (R)       36109                Mitt Romney    
1               . Barack Obama - (D)       14695               Barack Obama    
2               . Gary Johnson - (I)         381               Gary Johnson    
3               . Virgil Goode - (C)         117               Virgil Goode    
4                 . Jill Stein - (G)         113                 Jill Stein    
5   . Ross C. "Rocky" Anderson - (I)          43   Ross C. "Rocky" Anderson    
6              . Merlin Miller - (I)          41              Merlin Miller    
7                . Mitt Romney - (R)       69850                Mitt Romney    
8               . Barack Obama - (D)       25142               Barack Obama    
9               . Gary Johnson - (I)         748               Gary Johnson    
10                . Jill Stein - (G)         217                 Jill Stein    
11              . Virgil Goode - (C)         160               Virgil Goode    
12  . Ross C. "Rocky" Anderson - (I)          66   Ross C. "Rocky" Anderson    
13             . Merlin Miller - (I)          42              Merlin Miller    
14               . Mitt Romney - (R)        6197                Mitt Romney    
15              . Barack Obama - (D)        2795               Barack Obama    
16              . Gary Johnson - (I)          68               Gary Johnson    
17              . Virgil Goode - (C)          37               Virgil Goode    
18                . Jill Stein - (G)          22                 Jill Stein    
19  . Ross C. "Rocky" Anderson - (I)          11   Ross C. "Rocky" Anderson    

          party  
0    Republican  
1    Democratic  
2           (I)  
3           (C)  
4           (G)  
5           (I)  
6           (I)  
7    Republican  
8    Democratic  
9           (I)  
10          (G)  
11          (C)  
12          (I)  
13          (I)  
14   Republican  
15   Democratic  
16          (I)  
17          (C)  
18          (G)  
19          (I)

### Add column that has total vote count by count for use in calculating county vote percentages

In [42]:
# create new dataframe "county_vote" to use to find sum of votes per county
county_vote=tn_2012_pres.loc[0:,['county','vote_count']]
county_vote

county  vote_count
0      Wilson       36109
1      Wilson       14695
2      Wilson         381
3      Wilson         117
4      Wilson         113
..        ...         ...
659  Anderson         301
660  Anderson         101
661  Anderson          88
662  Anderson          54
663  Anderson          25

[664 rows x 2 columns]

In [43]:
#sum county votes
county_vote_sum=county_vote.groupby(['county'])['vote_count'].sum()
county_vote_sum

county
Anderson      29659
Bedford       14445
Benton         6226
Bledsoe        4359
Blount        49234
              ...  
Wayne          5486
Weakley       12337
White          9140
Williamson    96225
Wilson        51499
Name: vote_count, Length: 95, dtype: int64

In [44]:
# use "county_vote_sum" to add column "sum_county_vote" to tn_2016_pres correlating with value in each row
tn_2012_pres['sum_county_vote']=tn_2012_pres['county'].map(county_vote_sum)
tn_2012_pres.head(20)

electdate               officename      county  \
0   November 6, 2012  United States President      Wilson   
1   November 6, 2012  United States President      Wilson   
2   November 6, 2012  United States President      Wilson   
3   November 6, 2012  United States President      Wilson   
4   November 6, 2012  United States President      Wilson   
5   November 6, 2012  United States President      Wilson   
6   November 6, 2012  United States President      Wilson   
7   November 6, 2012  United States President  Williamson   
8   November 6, 2012  United States President  Williamson   
9   November 6, 2012  United States President  Williamson   
10  November 6, 2012  United States President  Williamson   
11  November 6, 2012  United States President  Williamson   
12  November 6, 2012  United States President  Williamson   
13  November 6, 2012  United States President  Williamson   
14  November 6, 2012  United States President       White   
15  November 6, 2012  United States President       White   
16  November 6, 2012  United States President       White   
17  November 6, 2012  United States President       White   
18  November 6, 2012  United States President       White   
19  November 6, 2012  United States President       White   

                           candidate  vote_count                        name  \
0                . Mitt Romney - (R)       36109                Mitt Romney    
1               . Barack Obama - (D)       14695               Barack Obama    
2               . Gary Johnson - (I)         381               Gary Johnson    
3               . Virgil Goode - (C)         117               Virgil Goode    
4                 . Jill Stein - (G)         113                 Jill Stein    
5   . Ross C. "Rocky" Anderson - (I)          43   Ross C. "Rocky" Anderson    
6              . Merlin Miller - (I)          41              Merlin Miller    
7                . Mitt Romney - (R)       69850                Mitt Romney    
8               . Barack Obama - (D)       25142               Barack Obama    
9               . Gary Johnson - (I)         748               Gary Johnson    
10                . Jill Stein - (G)         217                 Jill Stein    
11              . Virgil Goode - (C)         160               Virgil Goode    
12  . Ross C. "Rocky" Anderson - (I)          66   Ross C. "Rocky" Anderson    
13             . Merlin Miller - (I)          42              Merlin Miller    
14               . Mitt Romney - (R)        6197                Mitt Romney    
15              . Barack Obama - (D)        2795               Barack Obama    
16              . Gary Johnson - (I)          68               Gary Johnson    
17              . Virgil Goode - (C)          37               Virgil Goode    
18                . Jill Stein - (G)          22                 Jill Stein    
19  . Ross C. "Rocky" Anderson - (I)          11   Ross C. "Rocky" Anderson    

          party  sum_county_vote  
0    Republican            51499  
1    Democratic            51499  
2           (I)            51499  
3           (C)            51499  
4           (G)            51499  
5           (I)            51499  
6           (I)            51499  
7    Republican            96225  
8    Democratic            96225  
9           (I)            96225  
10          (G)            96225  
11          (C)            96225  
12          (I)            96225  
13          (I)            96225  
14   Republican             9140  
15   Democratic             9140  
16          (I)             9140  
17          (C)             9140  
18          (G)             9140  
19          (I)             9140

In [45]:
# create "pct_county_vote" column for tn_2016_pres
tn_2012_pres['pct_county_vote']=round(tn_2012_pres['vote_count']/tn_2012_pres['sum_county_vote']*100,2)
tn_2012_pres.head(20)

electdate               officename      county  \
0   November 6, 2012  United States President      Wilson   
1   November 6, 2012  United States President      Wilson   
2   November 6, 2012  United States President      Wilson   
3   November 6, 2012  United States President      Wilson   
4   November 6, 2012  United States President      Wilson   
5   November 6, 2012  United States President      Wilson   
6   November 6, 2012  United States President      Wilson   
7   November 6, 2012  United States President  Williamson   
8   November 6, 2012  United States President  Williamson   
9   November 6, 2012  United States President  Williamson   
10  November 6, 2012  United States President  Williamson   
11  November 6, 2012  United States President  Williamson   
12  November 6, 2012  United States President  Williamson   
13  November 6, 2012  United States President  Williamson   
14  November 6, 2012  United States President       White   
15  November 6, 2012  United States President       White   
16  November 6, 2012  United States President       White   
17  November 6, 2012  United States President       White   
18  November 6, 2012  United States President       White   
19  November 6, 2012  United States President       White   

                           candidate  vote_count                        name  \
0                . Mitt Romney - (R)       36109                Mitt Romney    
1               . Barack Obama - (D)       14695               Barack Obama    
2               . Gary Johnson - (I)         381               Gary Johnson    
3               . Virgil Goode - (C)         117               Virgil Goode    
4                 . Jill Stein - (G)         113                 Jill Stein    
5   . Ross C. "Rocky" Anderson - (I)          43   Ross C. "Rocky" Anderson    
6              . Merlin Miller - (I)          41              Merlin Miller    
7                . Mitt Romney - (R)       69850                Mitt Romney    
8               . Barack Obama - (D)       25142               Barack Obama    
9               . Gary Johnson - (I)         748               Gary Johnson    
10                . Jill Stein - (G)         217                 Jill Stein    
11              . Virgil Goode - (C)         160               Virgil Goode    
12  . Ross C. "Rocky" Anderson - (I)          66   Ross C. "Rocky" Anderson    
13             . Merlin Miller - (I)          42              Merlin Miller    
14               . Mitt Romney - (R)        6197                Mitt Romney    
15              . Barack Obama - (D)        2795               Barack Obama    
16              . Gary Johnson - (I)          68               Gary Johnson    
17              . Virgil Goode - (C)          37               Virgil Goode    
18                . Jill Stein - (G)          22                 Jill Stein    
19  . Ross C. "Rocky" Anderson - (I)          11   Ross C. "Rocky" Anderson    

          party  sum_county_vote  pct_county_vote  
0    Republican            51499            70.12  
1    Democratic            51499            28.53  
2           (I)            51499             0.74  
3           (C)            51499             0.23  
4           (G)            51499             0.22  
5           (I)            51499             0.08  
6           (I)            51499             0.08  
7    Republican            96225            72.59  
8    Democratic            96225            26.13  
9           (I)            96225             0.78  
10          (G)            96225             0.23  
11          (C)            96225             0.17  
12          (I)            96225             0.07  
13          (I)            96225             0.04  
14   Republican             9140            67.80  
15   Democratic             9140            30.58  
16          (I)             9140             0.74  
17          (C)             9140             0.40  
18          (G)             9140             0.24  
19      

In [46]:
# how many parties?
tn_2012_pres['party'].value_counts()

 (I)           284
 Republican     95
 Democratic     95
 (G)            95
 (C)            95
Name: party, dtype: int64

## Create dataframe "tn_2012_county_pres_pct"  with percent values per party per county

In [47]:
# create new dataframe with reduced columns values only
tn_2012_pct_county=tn_2012_pres.loc[0:,['electdate','county','party','pct_county_vote']]
tn_2012_pct_county

electdate    county        party  pct_county_vote
0    November 6, 2012    Wilson   Republican            70.12
1    November 6, 2012    Wilson   Democratic            28.53
2    November 6, 2012    Wilson          (I)             0.74
3    November 6, 2012    Wilson          (C)             0.23
4    November 6, 2012    Wilson          (G)             0.22
..                ...       ...          ...              ...
659  November 6, 2012  Anderson          (I)             1.01
660  November 6, 2012  Anderson          (C)             0.34
661  November 6, 2012  Anderson          (G)             0.30
662  November 6, 2012  Anderson          (I)             0.18
663  November 6, 2012  Anderson          (I)             0.08

[664 rows x 4 columns]

In [48]:
# create new dataframe with Republican values only
tn_2012_rep_pct=tn_2012_pct_county[tn_2012_pct_county['party']== " Republican"].reset_index(drop=True)
tn_2012_rep_pct

electdate      county        party  pct_county_vote
0   November 6, 2012      Wilson   Republican            70.12
1   November 6, 2012  Williamson   Republican            72.59
2   November 6, 2012       White   Republican            67.80
3   November 6, 2012     Weakley   Republican            69.75
4   November 6, 2012       Wayne   Republican            77.52
..               ...         ...          ...              ...
90  November 6, 2012      Blount   Republican            71.98
91  November 6, 2012     Bledsoe   Republican            69.33
92  November 6, 2012      Benton   Republican            61.84
93  November 6, 2012     Bedford   Republican            69.46
94  November 6, 2012    Anderson   Republican            63.95

[95 rows x 4 columns]

In [49]:
# rename 'pct_county_vote' to 'rep_pct' and drop "party" column to prepare for merge 
tn_2012_rep_pct=tn_2012_rep_pct.rename(columns={'pct_county_vote':'Republican'},inplace=False).drop(['party'], axis=1)
tn_2012_rep_pct

electdate      county  Republican
0   November 6, 2012      Wilson       70.12
1   November 6, 2012  Williamson       72.59
2   November 6, 2012       White       67.80
3   November 6, 2012     Weakley       69.75
4   November 6, 2012       Wayne       77.52
..               ...         ...         ...
90  November 6, 2012      Blount       71.98
91  November 6, 2012     Bledsoe       69.33
92  November 6, 2012      Benton       61.84
93  November 6, 2012     Bedford       69.46
94  November 6, 2012    Anderson       63.95

[95 rows x 3 columns]

In [50]:
# create new dataframe with Democratic values only
tn_2012_dem_pct=tn_2012_pct_county[tn_2012_pct_county['party']== " Democratic"].reset_index(drop=True)
tn_2012_dem_pct

electdate      county        party  pct_county_vote
0   November 6, 2012      Wilson   Democratic            28.53
1   November 6, 2012  Williamson   Democratic            26.13
2   November 6, 2012       White   Democratic            30.58
3   November 6, 2012     Weakley   Democratic            28.76
4   November 6, 2012       Wayne   Democratic            21.20
..               ...         ...          ...              ...
90  November 6, 2012      Blount   Democratic            26.27
91  November 6, 2012     Bledsoe   Democratic            29.07
92  November 6, 2012      Benton   Democratic            36.27
93  November 6, 2012     Bedford   Democratic            29.15
94  November 6, 2012    Anderson   Democratic            34.13

[95 rows x 4 columns]

In [51]:
# rename 'pct_county_vote' to 'dem_pct' and drop "party" to prepare for merge 
tn_2012_dem_pct=tn_2012_dem_pct.rename(columns={'pct_county_vote':'Democratic'},inplace=False).drop(['party'], axis=1)
tn_2012_dem_pct

electdate      county  Democratic
0   November 6, 2012      Wilson       28.53
1   November 6, 2012  Williamson       26.13
2   November 6, 2012       White       30.58
3   November 6, 2012     Weakley       28.76
4   November 6, 2012       Wayne       21.20
..               ...         ...         ...
90  November 6, 2012      Blount       26.27
91  November 6, 2012     Bledsoe       29.07
92  November 6, 2012      Benton       36.27
93  November 6, 2012     Bedford       29.15
94  November 6, 2012    Anderson       34.13

[95 rows x 3 columns]

In [52]:
# create new dataframe with non Republican or Democratic values only
tn_2012_other_pct=tn_2012_pct_county[(tn_2012_pct_county['party']!=" Republican")&(tn_2012_pct_county['party']!=" Democratic")].reset_index(drop=True)
tn_2012_other_pct

electdate    county party  pct_county_vote
0    November 6, 2012    Wilson   (I)             0.74
1    November 6, 2012    Wilson   (C)             0.23
2    November 6, 2012    Wilson   (G)             0.22
3    November 6, 2012    Wilson   (I)             0.08
4    November 6, 2012    Wilson   (I)             0.08
..                ...       ...   ...              ...
469  November 6, 2012  Anderson   (I)             1.01
470  November 6, 2012  Anderson   (C)             0.34
471  November 6, 2012  Anderson   (G)             0.30
472  November 6, 2012  Anderson   (I)             0.18
473  November 6, 2012  Anderson   (I)             0.08

[474 rows x 4 columns]

In [53]:
# sum by county
tn_2012_other_pct=tn_2012_other_pct.groupby(['county'],as_index=False)['pct_county_vote'].sum()
tn_2012_other_pct

county  pct_county_vote
0     Anderson             1.91
1      Bedford             1.38
2       Benton             1.89
3      Bledsoe             1.60
4       Blount             1.74
..         ...              ...
90       Wayne             1.28
91     Weakley             1.49
92       White             1.61
93  Williamson             1.29
94      Wilson             1.35

[95 rows x 2 columns]

In [54]:
# rename 'pct_county_vote' to 'other_pct' and drop "party" to prepare for merge 
tn_2012_other_pct=tn_2012_other_pct.rename(columns={'pct_county_vote':'Other'})
tn_2012_other_pct

county  Other
0     Anderson   1.91
1      Bedford   1.38
2       Benton   1.89
3      Bledsoe   1.60
4       Blount   1.74
..         ...    ...
90       Wayne   1.28
91     Weakley   1.49
92       White   1.61
93  Williamson   1.29
94      Wilson   1.35

[95 rows x 2 columns]

In [55]:
# join tn_2012_rep, tn_2012_dem_pct and tn_2012_other_pct to form tn_2012_county_pres_pct
dem=tn_2012_dem_pct['Democratic']
other=tn_2012_other_pct['Other']
tn_2012_county_pres_pct=tn_2012_rep_pct.join(dem).join(other)
tn_2012_county_pres_pct

electdate      county  Republican  Democratic  Other
0   November 6, 2012      Wilson       70.12       28.53   1.91
1   November 6, 2012  Williamson       72.59       26.13   1.38
2   November 6, 2012       White       67.80       30.58   1.89
3   November 6, 2012     Weakley       69.75       28.76   1.60
4   November 6, 2012       Wayne       77.52       21.20   1.74
..               ...         ...         ...         ...    ...
90  November 6, 2012      Blount       71.98       26.27   1.28
91  November 6, 2012     Bledsoe       69.33       29.07   1.49
92  November 6, 2012      Benton       61.84       36.27   1.61
93  November 6, 2012     Bedford       69.46       29.15   1.29
94  November 6, 2012    Anderson       63.95       34.13   1.35

[95 rows x 5 columns]

### Add column for Republican margin over Democrat for heat mapping in Tableau

In [56]:
tn_2012_county_pres_pct['Rep_Margin']= tn_2012_county_pres_pct['Republican']-tn_2012_county_pres_pct['Democratic']
tn_2012_county_pres_pct.head(50)

electdate      county  Republican  Democratic  Other  Rep_Margin
0   November 6, 2012      Wilson       70.12       28.53   1.91       41.59
1   November 6, 2012  Williamson       72.59       26.13   1.38       46.46
2   November 6, 2012       White       67.80       30.58   1.89       37.22
3   November 6, 2012     Weakley       69.75       28.76   1.60       40.99
4   November 6, 2012       Wayne       77.52       21.20   1.74       56.32
5   November 6, 2012  Washington       68.30       29.82   1.40       38.48
6   November 6, 2012      Warren       61.54       36.51   1.40       25.03
7   November 6, 2012   Van Buren       60.26       38.04   2.00       22.22
8   November 6, 2012       Union       73.47       25.36   1.39       48.11
9   November 6, 2012      Unicoi       71.01       27.00   1.58       44.01
10  November 6, 2012   Trousdale       55.49       42.69   1.68       12.80
11  November 6, 2012      Tipton       69.23       29.62   1.58       39.61
12  November 6, 2012      Sumner       70.38       28.42   1.27       41.96
13  November 6, 2012    Sullivan       72.74       25.58   1.28       47.16
14  November 6, 2012     Stewart       57.93       40.45   1.66       17.48
15  November 6, 2012       Smith       63.43       34.85   1.91       28.58
16  November 6, 2012      Shelby       36.55       62.63   0.83      -26.08
17  November 6, 2012      Sevier       76.73       21.91   1.32       54.82
18  November 6, 2012  Sequatchie       68.86       28.96   1.69       39.90
19  November 6, 2012       Scott       76.74       21.78   1.75       54.96
20  November 6, 2012  Rutherford       61.56       36.84   1.64       24.72
21  November 6, 2012   Robertson       67.11       31.53   1.71       35.58
22  November 6, 2012       Roane       69.82       28.53   0.89       41.29
23  November 6, 2012        Rhea       73.67       24.82   1.01       48.85
24  November 6, 2012      Putnam       67.66       30.60   1.33       37.06
25  November 6, 2012        Polk       67.80       30.63   1.57       37.17
26  November 6, 2012     Pickett       69.79       29.03   1.12       40.76
27  November 6, 2012       Perry       60.21       37.85   1.15       22.36
28  November 6, 2012     Overton       62.30       36.60   1.57       25.70
29  November 6, 2012       Obion       71.68       27.01   1.75       44.67
30  November 6, 2012      Morgan       71.79       26.52   1.85       45.27
31  November 6, 2012       Moore       73.35       25.19   1.37       48.16
32  November 6, 2012  Montgomery       54.41       44.08   1.70       10.33
33  November 6, 2012      Monroe       71.80       26.76   2.15       45.04
34  November 6, 2012       Meigs       68.97       29.34   0.88       39.63
35  November 6, 2012     Mcnairy       71.57       26.98   1.36       44.59
36  November 6, 2012      Mcminn       72.71       25.84   1.66       46.87
37  November 6, 2012       Maury       62.74       35.83   0.52       26.91
38  November 6, 2012    Marshall       63.61       34.68   1.17       28.93
39  November 6, 2012      Marion       60.26       37.98   1.62       22.28
40  November 6, 2012     Madison       54.03       45.13   1.92        8.90
41  November 6, 2012       Macon       76.18       22.48   1.58       53.70
42  November 6, 2012      Loudon       75.69       22.91   1.82       52.78
43  November 6, 2012     Lincoln       73.88       24.80   1.49       49.08
44  November 6, 2012       Lewis       66.40       30.83   1.65       35.57
45  November 6, 2012    Lawrence       70.77       27.84   1.61       42.93
46  November 6, 2012  Lauderdale       53.12       46.16   1.97        6.96
47  November 6, 2012        Lake       55.73       42.36   1.91       13.37
48  November 6, 2012        Knox       63.60       34.43   0.71       29.17
49  November 6, 2012     Johnson       74.44       23.94   1.40       50.50

In [57]:
# correct county names: Dekalb to DeKalb, Mcminn to McMinn, Mcnairy to McNairy 
tn_2012_county_pres_pct['county']=tn_2012_county_pres_pct['county'].str.replace('Dekalb','DeKalb').str.replace('Mcminn','McMinn').str.replace('Mcnairy','McNairy')
tn_2012_county_pres_pct.head(50)

electdate      county  Republican  Democratic  Other  Rep_Margin
0   November 6, 2012      Wilson       70.12       28.53   1.91       41.59
1   November 6, 2012  Williamson       72.59       26.13   1.38       46.46
2   November 6, 2012       White       67.80       30.58   1.89       37.22
3   November 6, 2012     Weakley       69.75       28.76   1.60       40.99
4   November 6, 2012       Wayne       77.52       21.20   1.74       56.32
5   November 6, 2012  Washington       68.30       29.82   1.40       38.48
6   November 6, 2012      Warren       61.54       36.51   1.40       25.03
7   November 6, 2012   Van Buren       60.26       38.04   2.00       22.22
8   November 6, 2012       Union       73.47       25.36   1.39       48.11
9   November 6, 2012      Unicoi       71.01       27.00   1.58       44.01
10  November 6, 2012   Trousdale       55.49       42.69   1.68       12.80
11  November 6, 2012      Tipton       69.23       29.62   1.58       39.61
12  November 6, 2012      Sumner       70.38       28.42   1.27       41.96
13  November 6, 2012    Sullivan       72.74       25.58   1.28       47.16
14  November 6, 2012     Stewart       57.93       40.45   1.66       17.48
15  November 6, 2012       Smith       63.43       34.85   1.91       28.58
16  November 6, 2012      Shelby       36.55       62.63   0.83      -26.08
17  November 6, 2012      Sevier       76.73       21.91   1.32       54.82
18  November 6, 2012  Sequatchie       68.86       28.96   1.69       39.90
19  November 6, 2012       Scott       76.74       21.78   1.75       54.96
20  November 6, 2012  Rutherford       61.56       36.84   1.64       24.72
21  November 6, 2012   Robertson       67.11       31.53   1.71       35.58
22  November 6, 2012       Roane       69.82       28.53   0.89       41.29
23  November 6, 2012        Rhea       73.67       24.82   1.01       48.85
24  November 6, 2012      Putnam       67.66       30.60   1.33       37.06
25  November 6, 2012        Polk       67.80       30.63   1.57       37.17
26  November 6, 2012     Pickett       69.79       29.03   1.12       40.76
27  November 6, 2012       Perry       60.21       37.85   1.15       22.36
28  November 6, 2012     Overton       62.30       36.60   1.57       25.70
29  November 6, 2012       Obion       71.68       27.01   1.75       44.67
30  November 6, 2012      Morgan       71.79       26.52   1.85       45.27
31  November 6, 2012       Moore       73.35       25.19   1.37       48.16
32  November 6, 2012  Montgomery       54.41       44.08   1.70       10.33
33  November 6, 2012      Monroe       71.80       26.76   2.15       45.04
34  November 6, 2012       Meigs       68.97       29.34   0.88       39.63
35  November 6, 2012     McNairy       71.57       26.98   1.36       44.59
36  November 6, 2012      McMinn       72.71       25.84   1.66       46.87
37  November 6, 2012       Maury       62.74       35.83   0.52       26.91
38  November 6, 2012    Marshall       63.61       34.68   1.17       28.93
39  November 6, 2012      Marion       60.26       37.98   1.62       22.28
40  November 6, 2012     Madison       54.03       45.13   1.92        8.90
41  November 6, 2012       Macon       76.18       22.48   1.58       53.70
42  November 6, 2012      Loudon       75.69       22.91   1.82       52.78
43  November 6, 2012     Lincoln       73.88       24.80   1.49       49.08
44  November 6, 2012       Lewis       66.40       30.83   1.65       35.57
45  November 6, 2012    Lawrence       70.77       27.84   1.61       42.93
46  November 6, 2012  Lauderdale       53.12       46.16   1.97        6.96
47  November 6, 2012        Lake       55.73       42.36   1.91       13.37
48  November 6, 2012        Knox       63.60       34.43   0.71       29.17
49  November 6, 2012     Johnson       74.44       23.94   1.40       50.50

In [58]:
#read TN_county_index_shape.csv to add index from TN Counties shape file for joining in Tableau
county_index=pd.read_csv('../data/TN_county_index_shape.csv')
county_index

county  INDEX_NO
0     Anderson        79
1      Bedford        26
2       Benton        41
3      Bledsoe        31
4       Blount        52
..         ...       ...
90       Wayne         6
91     Weakley        56
92       White        67
93  Williamson        44
94      Wilson        65

[95 rows x 2 columns]

In [59]:
# join with tn_2012_county_pres_pct
tn_2012_county_pres_pct=tn_2012_county_pres_pct.merge(county_index,on="county")
tn_2012_county_pres_pct

electdate      county  Republican  Democratic  Other  Rep_Margin  \
0   November 6, 2012      Wilson       70.12       28.53   1.91       41.59   
1   November 6, 2012  Williamson       72.59       26.13   1.38       46.46   
2   November 6, 2012       White       67.80       30.58   1.89       37.22   
3   November 6, 2012     Weakley       69.75       28.76   1.60       40.99   
4   November 6, 2012       Wayne       77.52       21.20   1.74       56.32   
..               ...         ...         ...         ...    ...         ...   
90  November 6, 2012      Blount       71.98       26.27   1.28       45.71   
91  November 6, 2012     Bledsoe       69.33       29.07   1.49       40.26   
92  November 6, 2012      Benton       61.84       36.27   1.61       25.57   
93  November 6, 2012     Bedford       69.46       29.15   1.29       40.31   
94  November 6, 2012    Anderson       63.95       34.13   1.35       29.82   

    INDEX_NO  
0         65  
1         44  
2         67  
3         56  
4          6  
..       ...  
90        52  
91        31  
92        41  
93        26  
94        79  

[95 rows x 7 columns]

In [60]:
tn_2012_county_pres_pct.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 95 entries, 0 to 94
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   electdate   95 non-null     object 
 1   county      95 non-null     object 
 2   Republican  95 non-null     float64
 3   Democratic  95 non-null     float64
 4   Other       95 non-null     float64
 5   Rep_Margin  95 non-null     float64
 6   INDEX_NO    95 non-null     int64  
dtypes: float64(4), int64(1), object(2)
memory usage: 5.9+ KB


In [61]:
# covert "electdate" to datetime from object
tn_2012_county_pres_pct['electdate']=pd.to_datetime(tn_2012_county_pres_pct['electdate'],format='%B %d, %Y')
tn_2012_county_pres_pct

electdate      county  Republican  Democratic  Other  Rep_Margin  INDEX_NO
0  2012-11-06      Wilson       70.12       28.53   1.91       41.59        65
1  2012-11-06  Williamson       72.59       26.13   1.38       46.46        44
2  2012-11-06       White       67.80       30.58   1.89       37.22        67
3  2012-11-06     Weakley       69.75       28.76   1.60       40.99        56
4  2012-11-06       Wayne       77.52       21.20   1.74       56.32         6
..        ...         ...         ...         ...    ...         ...       ...
90 2012-11-06      Blount       71.98       26.27   1.28       45.71        52
91 2012-11-06     Bledsoe       69.33       29.07   1.49       40.26        31
92 2012-11-06      Benton       61.84       36.27   1.61       25.57        41
93 2012-11-06     Bedford       69.46       29.15   1.29       40.31        26
94 2012-11-06    Anderson       63.95       34.13   1.35       29.82        79

[95 rows x 7 columns]

In [62]:
tn_2012_county_pres_pct.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 95 entries, 0 to 94
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   electdate   95 non-null     datetime64[ns]
 1   county      95 non-null     object        
 2   Republican  95 non-null     float64       
 3   Democratic  95 non-null     float64       
 4   Other       95 non-null     float64       
 5   Rep_Margin  95 non-null     float64       
 6   INDEX_NO    95 non-null     int64         
dtypes: datetime64[ns](1), float64(4), int64(1), object(1)
memory usage: 5.9+ KB


In [64]:
tn_2012_county_pres_pct.to_excel(r'../data/2012_election/tn_2012_county_pres_pct.xlsx',index=False)